source/opt/remove_duplicates_pass.cpp

```diff
@@ -0,0 +1,225 @@
+                             typesMap[b.GetSingleWordInOperand(0u)],
+                             typesMap) &&
+             a.GetSingleWordInOperand(1u) == b.GetSingleWordInOperand(1u);
+    case SpvOpTypeImage:
+      return areTypesSimilar(typesMap[a.GetSingleWordInOperand(0u)],
+                             typesMap[b.GetSingleWordInOperand(0u)],
+                             typesMap) &&
+             a.GetSingleWordInOperand(1u) == b.GetSingleWordInOperand(1u) &&
+             a.GetSingleWordInOperand(2u) == b.GetSingleWordInOperand(2u) &&
+             a.GetSingleWordInOperand(3u) == b.GetSingleWordInOperand(3u) &&
+             a.GetSingleWordInOperand(4u) == b.GetSingleWordInOperand(4u) &&
+             a.GetSingleWordInOperand(5u) == b.GetSingleWordInOperand(5u) &&
+             a.GetSingleWordInOperand(6u) == b.GetSingleWordInOperand(6u) &&
+             a.NumOperands() == b.NumOperands() &&
+             (a.NumInOperands() == 8u ||
+              a.GetSingleWordInOperand(7u) == b.GetSingleWordInOperand(7u));
+    case SpvOpTypeSampledImage:
+    case SpvOpTypeRuntimeArray:
+      return areTypesSimilar(typesMap[a.GetSingleWordInOperand(0u)],
+                             typesMap[b.GetSingleWordInOperand(0u)], typesMap);
+    case SpvOpTypeArray:
+      return areTypesSimilar(typesMap[a.GetSingleWordInOperand(0u)],
+                             typesMap[b.GetSingleWordInOperand(0u)],
+                             typesMap) &&
+             areTypesSimilar(typesMap[a.GetSingleWordInOperand(1u)],
+                             typesMap[b.GetSingleWordInOperand(1u)], typesMap);
+    case SpvOpTypeStruct:
+    case SpvOpTypeFunction: {
+      bool res = a.NumInOperands() == b.NumInOperands();
+      for (uint32_t i = 0u; i < a.NumInOperands() && res; ++i)
+        res &= areTypesSimilar(typesMap[a.GetSingleWordInOperand(i)],
+                               typesMap[b.GetSingleWordInOperand(i)], typesMap);
+      return res;
+    }
+    case SpvOpTypeOpaque:
+      return std::strcmp(
+                 reinterpret_cast<const char*>(a.GetInOperand(0u).words.data()),
+                 reinterpret_cast<const char*>(
+                     b.GetInOperand(0u).words.data())) == 0;
+    case SpvOpTypePointer:
+      return a.GetSingleWordInOperand(0u) == b.GetSingleWordInOperand(0u) &&
+             areTypesSimilar(typesMap[a.GetSingleWordInOperand(1u)],
+                             typesMap[b.GetSingleWordInOperand(1u)], typesMap);
+    default:
+      return false;
+  }
+}
+
+}  // anonymous namespace
+
+namespace spvtools {
+namespace opt {
+
+using ir::Instruction;
+using ir::Operand;
+
+Pass::Status RemoveDuplicatesPass::Process(ir::Module* module) {
+  bool modified = false;
+
+  // Remove duplicate capabilities
+  std::unordered_set<uint32_t> capabilities;
+  for (auto i = module->capability_begin(); i != module->capability_end();) {
+    auto res = capabilities.insert(i->GetSingleWordOperand(0u));
+    i = (res.second) ? ++i : i.Erase();
```

Comment:
Wait, weren’t you the one that suggested I should use it, in one of your previous reviews? :-p I could very well be misremembering though.
I don’t mind at all switching back to a more readable form.